### PR TITLE
refactor(dev): add jsdocs

### PR DESCRIPTION
**Pull request by tverney**
## Problem

We miss adding support for JSDocs on amazonqFeatureDev based on the types.

Some benefits of this:

- Intelligent Code Completion: JSDoc comments provide information about functions, parameters, return types, and more. VS Code uses this information to offer intelligent code completion suggestions, making it easier to write code and understand the available methods and properties.

- Documentation on Hover: When you hover over a function or variable that has JSDoc comments, VS Code displays a tooltip with the documentation. This makes it easier to understand the purpose and usage of different parts of your code without having to navigate away from your current file.

Screenshot 2024-09-06 at 5 12 21 PM

- Type Checking and Error Detection: JSDoc comments can include type information, which allows VS Code's built-in TypeScript support to perform static type checking. This can help catch type-related errors early and improve the overall quality of your code.

- Code Navigation: JSDoc comments can include links to other parts of your codebase or external resources. VS Code can detect these links and allow you to navigate to them quickly, making it easier to explore and understand your code.

- Documentation Generation: Tools like JSDoc can automatically generate HTML documentation from your JSDoc comments. 

- Code Folding: VS Code can automatically fold or collapse code sections based on JSDoc comments, making it easier to navigate and organize your code.

- Better Developer Experience: Well-documented code with JSDoc comments makes it easier for developers to understand and work with the codebase. This can improve productivity, reduce onboarding time for new team members, and facilitate collaboration.

## Solution

This PR add those changes under /amazonqFeatureDev.

---

License: I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/core/src/amazonqFeatureDev/util/files.ts [hunk 1]
```diff
@@ -63,15 +70,18 @@ export async function prepareRepoData(
         const iterator = ignoredExtensionMap.entries()
 
         for (let i = 0; i < ignoredExtensionMap.size; i++) {
-            const [key, value] = iterator.next().value
-            await amznTelemetry.amazonq_bundleExtensionIgnored.run(async (bundleSpan) => {
-                const event = {
-                    filenameExt: key,
-                    count: value,
-                }
+            const nextValue = iterator.next().value
```
Comment (tverney) [inline packages/core/src/amazonqFeatureDev/util/files.ts:73]: Including this Typescript validation fix.

File: packages/core/src/amazonqFeatureDev/errors.ts [hunk 2]
```diff
@@ -26,12 +34,20 @@ export class TabIdNotFoundError extends ToolkitError {
...
+ */
 export class PanelLoadError extends ToolkitError {
     constructor() {
         super(`${featureName} UI panel failed to load`, { code: 'PanelLoadFailed' })
     }
 }
 
+/**
+ * Error thrown when the workspace folder is not found.
+ * @extends ToolkitError
```
Comment (justinmk3) [inline packages/core/src/amazonqFeatureDev/errors.ts:49]: why are we adding `@extends` jsdoc directives? this is a typescript project. there is already a `extends` declaration in the code below.

File: packages/core/src/amazonqFeatureDev/controllers/chat/messenger/messenger.ts [hunk 3]
```diff
@@ -24,8 +24,23 @@ import { FeatureAuthState } from '../../../../codewhisperer'
 import { CodeReference } from '../../../../codewhispererChat/view/connector/connector'
 import { i18n } from '../../../../shared/i18n-helper'
 export class Messenger {
+    /**
+     * Creates an instance of Messenger.
```
Comment (justinmk3) [inline packages/core/src/amazonqFeatureDev/controllers/chat/messenger/messenger.ts:28]: Were these docstrings generated by AI? Most of these are just restating the function name. That is counterproductive, it's actually worse to have extra text to read which merely duplicates the function name, compared to having no text at all.

Meaningful comments which give insight and hints, are *very* helpful. Meaningless comments are actively harmful.

**Review by justinmk3 — CHANGES_REQUESTED** [review]
Were these docstrings generated by AI? Most of these are just restating the function name. That is counterproductive, it's actually worse to have extra text to read which merely duplicates the function name, compared to having no text at all.

Meaningful comments which give insight and hints, are *very* helpful. Meaningless comments are actively harmful.

File: packages/core/src/amazonqFeatureDev/controllers/chat/messenger/messenger.ts (hunk 3 above)
Comment (tverney) [reply packages/core/src/amazonqFeatureDev/controllers/chat/messenger/messenger.ts:28]: Sorry @justinmk3, this is on draft, as you can see. Isn't on my intent to publish this now.

**Comment by ivikash** [comment]
Can I use `/dev` to automatically generate JSDoc for the entire codebase?

**Comment by tverney** [comment]
> Can I use `/dev` to automatically generate JSDoc for the entire codebase?

that's what I've used :)

**Comment by ivikash** [comment]
> > Can I use `/dev` to automatically generate JSDoc for the entire codebase?
> 
> that's what I've used :)

Awesome! We can probably create showcase this as a blog on community.aws.

**Pull request closed without merging**